### PR TITLE
Edge case fixes

### PR DESCRIPTION
**Pull request by TroyDota**
Pull request checklist:

- [ ] `CHANGELOG.md` was updated, if applicable

### CODE REVIEW

File: pajbot/managers/discord_bot.py [hunk 1]
```diff
@@ -73,11 +73,11 @@ def __init__(self, bot, redis):
         self.discord_task = self.schedule_task_periodically(300, self.check_discord_roles)
         queued_subs = self.redis.get("queued-subs-discord")
         unlinkinfo = self.redis.get("unlinks-subs-discord")
-        if unlinkinfo is None:
-            data = {"array": []}
+        if unlinkinfo is None or "array" in json.loads(unlinkinfo):
```
Comment (TroyDota) [inline pajbot/managers/discord_bot.py]: this is to allow for easy migration

File: pajbot/managers/discord_bot.py [hunk 2]
```diff
@@ -73,11 +73,11 @@ def __init__(self, bot, redis):
...
+            data = {}
             self.redis.set("unlinks-subs-discord", json.dumps(data))
-        if queued_subs is None:
-            data = {"array": []}
+        if queued_subs is None or "array" in json.loads(queued_subs):
```
Comment (TroyDota) [inline pajbot/managers/discord_bot.py]: this is to allow for easy migration

File: pajbot/managers/discord_bot.py [hunk 3]
```diff
@@ -92,6 +92,15 @@ def add_command(self, *args, **kwargs):
                 await self.private_message(message.author, f"Check Complete!")
                 return
 
+    def get_discord_string(self, id):
```
Comment (TroyDota) [inline pajbot/managers/discord_bot.py]: Edge cases where they left the guild or left discord

File: pajbot/managers/discord_bot.py [hunk 4]
```diff
@@ -160,183 +169,178 @@ def add_command(self, *args, **kwargs):
             tier3_role = self.guild.get_role(int(self.settings["tier3_role"]))
             notify_role = self.guild.get_role(int(self.settings["notify_role"]))
             ignore_role = self.guild.get_role(int(self.settings["ignore_role"]))
+            roles_allocated = {
```
Comment (TroyDota) [inline pajbot/managers/discord_bot.py]: Allows for super compression

File: pajbot/managers/discord_bot.py [hunk 5]
```diff
@@ -160,183 +169,178 @@ def add_command(self, *args, **kwargs):
...
+            if twitch_sub_role is None:
+                return
             quick_dict = {}
             with DBManager.create_session_scope() as db_session:
                 all_connections = db_session.query(UserConnections).all()
                 for connection in all_connections:
                     user_linked = User.find_by_id(db_session, connection.twitch_id)
-                    quick_dict[connection.discord_user_id] = [user_linked.tier, connection]
                     member = self.guild.get_member(int(connection.discord_user_id))
+                    if not user_linked or (
```
Comment (TroyDota) [inline pajbot/managers/discord_bot.py]: Allows for database pruning

File: pajbot/web/routes/base/user.py [hunk 6]
```diff
@@ -97,8 +97,10 @@ def user_profile(login):
                     "winrate_str": f"{winrate * 100:.2f}%",
                     "roulette_base_winrate": roulette_base_winrate,
                 }
-
-            return render_template("user.html", user=user, roulette_stats=roulette_stats, roulettes=roulettes)
+            paired = True if UserConnections._from_twitch_id(db_session, user.id) else False
```
Comment (DatGuy1) [inline pajbot/web/routes/base/user.py]: bool(UserConnections.xyz) should work, test it

File: pajbot/managers/discord_bot.py [hunk 7]
```diff
@@ -92,13 +93,63 @@ def add_command(self, *args, **kwargs):
                 await self.private_message(message.author, f"Check Complete!")
                 return
 
+    async def _get_users_by_tier(self, message):
+        if self.guild:
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: if not self.guild: return

Helps with indentation

File: pajbot/managers/discord_bot.py [hunk 8]
```diff
@@ -92,13 +93,63 @@ def add_command(self, *args, **kwargs):
...
+        id = int(id)
+        member = self.guild.get_member(id) or self.client.get_user(id)
+        return (
+            f"\nDiscord: {member.display_name}#{member.discriminator} (<https://discordapp.com/users/{member.id}>)"
+            if member
+            else ""
+        )
+
     async def _connections(self, message):
         if self.guild:
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: Same here and all other occurrences

File: pajbot/managers/discord_bot.py [hunk 9]
```diff
@@ -160,183 +213,188 @@ def add_command(self, *args, **kwargs):
...
-                    for member in tier3_role.members:
-                        if ignore_role is None or ignore_role not in member.roles:
-                            if twitch_sub_role not in member.roles or str(member.id) not in quick_dict:
-                                await self.remove_role(member, tier3_role)
-                data = {"array": subs_to_return}
-                self.redis.set("queued-subs-discord", json.dumps(data))
-                data = {"array": []}
-                self.redis.set("unlinks-subs-discord", json.dumps(data))
-                db_session.commit()
+                for tier in range(2, 4):
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: C:\Users\Acer>python -m timeit "for i in range(2, 4): xd = i"
1000000 loops, best of 5: 306 nsec per loop

C:\Users\Acer>python -m timeit "for i in [2, 3]: xd = i"
5000000 loops, best of 5: 85 nsec per loop

File: pajbot/managers/discord_bot.py [hunk 10]
```diff
@@ -160,183 +213,188 @@ def add_command(self, *args, **kwargs):
...
+                            if ignore_role is None or ignore_role not in member.roles:
+                                connection = UserConnections._from_discord_id(db_session, str(member.id))
+                                if not connection:
+                                    await self.remove_role(member, role)
+                                elif connection.tier == tier:
+                                    continue
+                                elif connection.twitch_id not in subs_to_return:
+                                    if connection.tier != 0:
+                                        await self.remove_role(member, role)
+                                    else:
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: elif

**Comment by TroyDota** [comment]
I'll fix this as soon as I get home.

File: pajbot/managers/discord_bot.py (hunk 10 above)
Comment (TroyDota) [reply pajbot/managers/discord_bot.py]: What is wrong?

File: pajbot/managers/discord_bot.py (hunk 9 above)
Comment (TroyDota) [reply pajbot/managers/discord_bot.py]: Will fix

File: pajbot/managers/discord_bot.py (hunk 7 above)
Comment (TroyDota) [reply pajbot/managers/discord_bot.py]: Will fix

File: pajbot/web/routes/base/user.py (hunk 6 above)
Comment (TroyDota) [reply pajbot/web/routes/base/user.py]: Will test and update if true

**Pull request merged**